src/plugins/data/public/ui/dataset_selector/dataset_selector.tsx

```diff
@@ -99,22 +99,41 @@ export const DatasetSelector = ({
         prepend: <EuiIcon type={datasetService.getType(type)!.meta.icon.type} />,
       });
     });
-
+    const recentlySelected: EuiSelectableOption[] = [
+      {
+        label: 'Recently Selected Data',
+        isGroupLabel: true,
+      },
+    ];
+    datasetService.getRecentDatasets().forEach(({ id, title, type, dataSource }) => {
+      const label = dataSource ? `${dataSource.title}::${title}` : title;
+      recentlySelected.push({
+        label,
+        checked: id === selectedDataset?.id ? 'on' : undefined,
+        key: id,
+        prepend: <EuiIcon type={datasetService.getType(type)!.meta.icon.type} />,
+      });
+    });
+    if (recentlySelected.length > 1) {
+      newOptions.unshift(...recentlySelected);
+    }
     return newOptions;
   }, [datasets, selectedDataset?.id, datasetService]);
 
   const handleOptionChange = useCallback(
     (newOptions: EuiSelectableOption[]) => {
       const selectedOption = newOptions.find((option) => option.checked === 'on');
       if (selectedOption) {
-        const foundDataset = datasets.find((dataset) => dataset.id === selectedOption.key);
+        const foundDataset =
+          datasets.find((dataset) => dataset.id === selectedOption.key) ??
+          datasetService.getRecentDataset(selectedOption.key);
```

Comment:
Cant the recents be added along with the fetched datasets so that we have a single source of truth and avoids adding another dependency to the useCallback hook?

Reply:
+1

Reply:
The `datasets` state in this file is fetched using `fetchDatasets()`, which just calls the `fetch()` function for index patterns. Since it is just an array of index patterns, we would need to identify which datasets in the array belong to the index pattern section and the recently selected datasets section. I separated them here because it's possible that the recent datasets are not index patterns. To use a single source of truth, I think the easiest way would be to build the options array in `fetchDatasets()` such that the `datasets` state already contains the labels differentiating the two sections in the array. Open to other suggestions though.